manager/allocator/allocator_new.go

```diff
@@ -0,0 +1,929 @@
+
+		batchDone := metrics.StartTimer(storeLockHeld.WithValues("networks", "allocate", "read"))
+		a.store.View(func(tx store.ReadTx) {
+			for id := range pendingNetworks {
+				// remove the network from the pending allocations. if something
+				// fails and we need to retry, we'll add it back.
+				nw := store.GetNetwork(tx, id)
+				if nw == nil {
+					log.G(ctx).WithField("network.id", id).Debug("network not found, probably deleted")
+					continue
+				}
+				networks = append(networks, nw)
+			}
+		})
+		batchDone()
+
+		// now, allocate each network
+		for _, network := range networks {
+			ctx := log.WithField(ctx, "network.id", network.ID)
+			allocDone := metrics.StartTimer(allocatorActions.WithValues("network", "allocate"))
+			err := a.network.AllocateNetwork(network)
+			allocDone()
+
+			// error conditions from the allocator are numerous, but are
+			// constrained to a few specific types, which we can completely cover.
+			switch {
+			case err == nil:
+				// first, if no error occurred, then we can add this to the list of
+				// networks we will commit
+				allocatedNetworks = append(allocatedNetworks, network)
+			case errors.IsErrAlreadyAllocated(err):
+				// if the network is already allocated, there is nothing to do. log
+				// that and do not add this network to the batch
+				log.G(ctx).Debug("network already fully allocated")
+			case errors.IsErrRetryable(err):
+				// if the error is retryable, then we should log that error and
+				// re-add this network to our pending networks
+				log.G(ctx).WithError(err).Error("network could not be allocated, but will be retried")
+				a.pendingNetworks[network.ID] = struct{}{}
+			default:
+				// default covers any other error case. specifically, error that
+				// are not retryable. for these, we should fail and not re-add them
+				// to the pending map. they will never succeed
+				log.G(ctx).WithError(err).Error("network cannot be allocated")
+			}
+		}
+
+		// again, if we haven't actually successfully allocated any networks,
+		// we should skip this part to avoid holding the lock
+		if len(allocatedNetworks) > 0 {
+			batchDone = metrics.StartTimer(storeLockHeld.WithValues("network", "allocate", "write"))
+			// now, batch update the networks we just allocated
+			if err := a.store.Batch(func(batch *store.Batch) error {
+				for _, network := range allocatedNetworks {
+					if err := batch.Update(func(tx store.Tx) error {
+						ctx := log.WithField(ctx, "network.id", network.ID)
+						// first, get the network and make sure it still exists
+						currentNetwork := store.GetNetwork(tx, network.ID)
+						if currentNetwork == nil {
+							log.G(ctx).WithField("network.id", network.ID).Debugf("network was deleted after allocation but before commit")
+							// if the current network is nil, then the network was
+							// deleted and we should deallocate it.
+							a.network.DeallocateNetwork(network)
```

Comment:
I guess `DeallocateNetwork` is idempotent and hence would not experience a similar race to `DeallocateNode`?  (looks like we check for the network ID and don't do anything if it doesn't exist).  It probably doesn't make sense to keep track of node IDs in the network allocator, since the node ID is not needed, but would that be one way of making the node deallocations idempotent?

Reply:
this is addressed, i think? i think this review comment happens to refer to an unchanged section of code.